tokio/src/net/windows/named_pipe.rs

```diff
@@ -0,0 +1,644 @@
+use winapi::shared::minwindef::DWORD;
+use winapi::um::fileapi;
+use winapi::um::handleapi;
+use winapi::um::minwinbase;
+use winapi::um::namedpipeapi;
+use winapi::um::winbase;
+use winapi::um::winnt;
+
+use crate::fs::asyncify;
+use crate::io::{AsyncRead, AsyncWrite, Interest, PollEvented, ReadBuf};
+
+// Interned constant to wait forever. Not available in winapi.
+const NMPWAIT_WAIT_FOREVER: DWORD = 0xffffffff;
+
+/// A [Windows named pipe].
+///
+/// [Windows named pipe]: https://docs.microsoft.com/en-us/windows/win32/ipc/named-pipes
+///
+/// Constructed using [NamedPipeClientBuilder::create] for clients, or
+/// [NamedPipeBuilder::create] for servers. See their corresponding
+/// documentation for examples.
+#[derive(Debug)]
+pub struct NamedPipe {
+    io: PollEvented<mio::windows::NamedPipe>,
+}
+
+impl NamedPipe {
+    /// Fallibly construct a new named pipe from the specified raw handle.
+    ///
+    /// This function will consume ownership of the handle given, passing
+    /// responsibility for closing the handle to the returned object.
+    ///
+    /// This function is also unsafe as the primitives currently returned have
+    /// the contract that they are the sole owner of the file descriptor they
+    /// are wrapping. Usage of this function could accidentally allow violating
+    /// this contract which can cause memory unsafety in code that relies on it
+    /// being true.
+    ///
+    /// # Errors
+    ///
+    /// This errors if called outside of a [Tokio Runtime] which doesn't have
+    /// [I/O enabled].
+    ///
+    /// [Tokio Runtime]: crate::runtime::Runtime
+    /// [I/O enabled]: crate::runtime::RuntimeBuilder::enable_io
+    unsafe fn try_from_raw_handle(handle: RawHandle) -> io::Result<Self> {
+        let named_pipe = mio::windows::NamedPipe::from_raw_handle(handle);
+
+        Ok(NamedPipe {
+            io: PollEvented::new(named_pipe)?,
+        })
+    }
+
+    /// Enables a named pipe server process to wait for a client process to
+    /// connect to an instance of a named pipe. A client process connects by
+    /// creating a named pipe with the same name.
+    ///
+    /// This corresponds to the [`ConnectNamedPipe`] system call.
+    ///
+    /// [`ConnectNamedPipe`]: https://docs.microsoft.com/en-us/windows/win32/api/namedpipeapi/nf-namedpipeapi-connectnamedpipe
+    ///
+    /// ```no_run
+    /// use tokio::net::windows::NamedPipeBuilder;
+    ///
+    /// # #[tokio::main] async fn main() -> std::io::Result<()> {
+    /// let builder = NamedPipeBuilder::new("\\\\.\\pipe\\mynamedpipe");
+    /// let pipe = builder.create()?;
+    ///
+    /// // Wait for a client to connect.
+    /// pipe.connect().await?;
+    ///
+    /// // Use the connected client...
+    /// # Ok(()) }
+    /// ```
+    pub async fn connect(&self) -> io::Result<()> {
+        loop {
+            match self.io.connect() {
+                Ok(()) => break,
+                Err(e) if e.kind() == io::ErrorKind::WouldBlock => {
```

Comment:
So with the current API, this means that the type system doesn't prevent you from calling this on a client socket, or on a connected socket. This might be ok, but it is worth to discuss.

Reply:
After giving the available APIs a more thorough look through I am loosening up to the idea of having `NamedPipeClient` and `NamedPipeServer`. There aren't that many common APIs, and the ones that exist (e.g. `GetNamedPipeInfo`, `TransactNamedPipe`, `PeekNamedPipe`, ...) could be duplicated if we find it necessary to introduce them. Still, the idea of only having one makes the API more obvious to design - but leaves it up to the user to use correctly.

It is worth noting that anonymous named pipes (as the ones created through `CreatePipe` <https://docs.microsoft.com/en-us/windows/win32/api/namedpipeapi/nf-namedpipeapi-createpipe>) would probably require their own wrapper, or maybe just use `NamedPipeClient`.

Reply:
The `miow` crate uses a separate type for the anonymous pipes.

Reply:
So after a bit of work I've come to the conclusion that `CreatePipe` et. al. does not appear to be suitable for constructing handles that can be used in Tokio. There's no way to set `FILE_FLAG_OVERLAPPED` when calling it.

So let's punt the introduction of anonymous pipes*, but that doesn't mean we should just ignore that they might be introduced in the future in how we decide to name things - which it's probably time to decide now.

*: building your own CreatePipe implementation [is quite straight forward](https://stackoverflow.com/questions/60645/overlapped-i-o-on-anonymous-pipe).